docs/porting/target/spi.md

```diff
@@ -0,0 +1,133 @@
+
+The **Serial Peripheral Interface** allows you to send or receive a data stream over a synchronous serial interface made of 3 to 4 lines.
+
+- MISO: Master In Slave Out.
+- MOSI: Master Out Slave In.
+- MCLK: Clock.
+- SS: Slave Select.
+
+A typical use case of this interface is with SDCard, memory blocks and DACs or ADCs.
+
+This highly configurable interface has a wide range of elements you can adjust:
+
+- Frame length.
+- Clocks polarity and phase.
+
+<span class="warnings">**Warning:** We are introducing the SPI API in an upcoming release of Mbed OS. This page documents code that exists on a feature branch of Mbed OS. You can find details on how it may affect you in the [implementing the SPI API](#implementing-the-spi-api) section.
+
+### Assumptions
+
+#### Defined Behaviours
+
+- `spi_get_module()` returns the `SPIName` unique identifier to the peripheral associated to this SPI channel.
+- `spi_get_capabilities()` fills the given `spi_capabilities_t` instance
+- `spi_get_capabilities()` should consider the `ssel` pin when evaluation the `support_slave_mode` capability.  
+  If the given `ssel` pin cannot be managed by hardware in slave mode, `support_slave_mode` should be false.
+- At least a symbol width of 8bit must be supported.
+- The supported frequency range must include the range [0.2..2] MHz.
+- The shortest part of the duty cycle must not be shorter than 50% of the expected period.
+- `spi_init()` initializes the pins leaving the configuration registers unchanged.
+- `spi_init()` if `is_slave` is false:
+    - if `ssel` is `NC` the hal implementation ignores this pin.
+    - if `ssel` is not `NC` then the hal implementation owns the pin and its management.
+- When managed by the hal implementation, `ssel` is always considered active low.
+- When the hardware supports the half-duplex (3-wire) mode, if `miso` (exclusive) or `mosi` is missing in any function that expects pins, the bus is assumed to be half-duplex.
+- `spi_free()` resets the pins to their default state.
+- `spi_free()` disables the peripheral clock.
+- `spi_format()` sets :
+  - the number of bits per symbol
+  - the mode :
+    0. Clock idle state is *low*, data are sampled when the clock becomes *active* (polarity = 0, phase = 0)
+    1. Clock idle state is *low*, data are sampled when the clock becomes *inactive* (polarity = 0, phase = 1)
+    2. Clock idle state is *high*, data are sampled when the clock becomes *active* (polarity = 1, phase = 0)
+    3. Clock idle state is *high*, data are sampled when the clock becomes *inactive* (polarity = 1, phase = 1)
+  - the bit ordering (lsb/msb first).
+- `spi_format()` updates the configuration of the peripheral except the baud rate generator.
+- `spi_frequency()` sets the frequency to use during the transfer.
+- `spi_frequency()` returns the actual frequency that will be used.
+- `spi_frequency()` updates the baud rate generator leaving other configurations unchanged.
```

Comment:
Why format() should leave baud rate generator un-touched? And the frequency() should only configure the baud rate generator? Can we have format() setting everything and frequency() change "only" the baud rate gen along with the other 2 functions above?

Reply:
Because the responsibility of `spi_format` is only to configure the format of the frames. Changing the baudrate generator in `spi_format` could be considered as a side effect as it overlaps with the dedicated method `spi_frequency`.

The choice of having two functions was made during the [RFC process](https://github.com/ARMmbed/mbed-os/pull/7671) because the format is very unlikely to change during the lifetime of a spi instance while the frequency is likely to be raised after the first few exchanges with the slave peripheral.